### PR TITLE
ENH: Add numeric_only to certain groupby ops

### LINKED ISSUE #46524: ENH: Add numeric_only to groupby frame ops
Once #46072 is implemented, many groupby ops will be defaulting to `numeric_only=False` in 2.0. However there are a number of group ops which can only ever work on numeric data. For API consistency, I believe a user trying to operate on non-numeric columns with these ops should raise. Consider the example

```
df = pd.DataFrame({'a': [1, 1], 'b': [3, 4], 'c': [5, 6]})
df['c'] = df['c'].astype(object)
gb = df.groupby('a')
print(gb.mean())
```

which gives the output

```
     b
a     
1  3.5
```

If a user has a numeric column that accidentally ends up as object dtype, the result will be silently missing expected columns. This is why I think we should run the op with all provided data, regardless if it is numeric or not.

The following groupby ops have no `numeric_only` argument and act like `numeric_only=True`, but only make sense on numeric data.

 - [ ] quantile #46728
 - [ ] std #46728
 - [x] mad #46707
 - [ ] sem #46728
 - [ ] var #46728
 - [x] corrwith #46708
 - [x] cumprod #46708
 - [x] cov #46708
 - [x] corr #46708

The following groupby ops have no `numeric_only` argument and act like `numeric_only=True`, but make sense on non-numeric data.

 - [ ] idxmin #46728
 - [ ] idxmax #46728
 - [x] cummax #46708
 - [x] cumsum #46708
 - [x] cummin #46708

For both groups of ops, I propose we add the numeric_only argument defaulting to True in 1.5, which emits a warning message that it will default to False in the future. The warning would only be emitted if setting numeric_only to True/False would give rise to different output; i.e. if there are non-numeric columns that could have been operated on.

It's not ideal to add an argument and deprecate the default value in the same minor release (assuming 1.5 is the last minor release in the 1.x series), however I believe it will be of minor impact to users. The alternatives would be not carrying out the deprecation of `numeric_only=True` or to leave these ops behaving as if `numeric_only=True` (with no numeric_only argument). Both of these seem like worse alternatives to me.

cc @jreback @jbrockmendel @jorisvandenbossche @simonjayhawkins @Dr-Irv

### PR DESCRIPTION
**Pull request by rhshadrach**
Part of #46560

- [x] Closes #46524  
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

For the ops that have numeric_only added, we have the option of dropping nuisance columns even when `numeric_only=False` is specified to be consistent with other groupby ops (e.g. sum). I decided to implement the correct behavior here instead; this required adding an argument to the cython/python paths. This argument can be removed in 2.0 when all ops would raise.

### CODE REVIEW

**Comment by rhshadrach** [comment]
@github-actions pre-commit

pre-commit check timed out.

File: pandas/core/groupby/generic.py [hunk 1]
```diff
@@ -1553,10 +1553,14 @@ def nunique(self, dropna: bool = True) -> DataFrame:
...
-    def idxmax(self, axis=0, skipna: bool = True):
+    @doc(
+        _shared_docs["idxmax"],
+        numeric_only_default="True for axis=0, False for axis=1",
+    )
+    def idxmax(self, axis=0, skipna: bool = True, numeric_only: bool | None = None):
         axis = DataFrame._get_axis_number(axis)
-        numeric_only = None if axis == 0 else False
+        if numeric_only is None:
+            numeric_only = None if axis == 0 else False
```
Comment (mroeschke) [inline pandas/core/groupby/generic.py:1563]: Do you think this `None/False` behavior should be deprecated and changed in the future too to match the other ops?

Comment (rhshadrach) [reply pandas/core/groupby/generic.py:1563]: Yep! This PR adds numeric_only to the necessary ops, I have two follow ons that deprecate for frame and groupby ops respectively that are just about ready.

**Review by mroeschke — APPROVED** [review]

**Comment by rhshadrach** [comment]
@jreback - gentle ping.

File: pandas/core/groupby/groupby.py [hunk 2]
```diff
@@ -3002,6 +3051,7 @@ def quantile(self, q=0.5, interpolation: str = "linear"):
         a    2.0
         b    3.0
         """
+        numeric_only_bool = self._resolve_numeric_only(numeric_only)
```
Comment (jreback) [inline pandas/core/groupby/groupby.py:3054]: why don't u use this in the other function (id min and so on)

File: pandas/core/groupby/groupby.py [hunk 3]
```diff
@@ -2133,22 +2153,37 @@ def var(
...
+                if numeric_only:
+                    nonnumeric_exclusions = frozenset(
+                        self.obj.columns.difference(self.exclusions).difference(
+                            self.obj._get_numeric_data().columns
+                        )
+                    )
+                else:
+                    nonnumeric_exclusions = frozenset()
+                with com.temp_setattr(
+                    self, "exclusions", self.exclusions | nonnumeric_exclusions
```
Comment (jreback) [inline pandas/core/groupby/groupby.py:2176]: what is the purpose here?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/groupby/groupby.py (hunk 2 above)
Comment (rhshadrach) [reply pandas/core/groupby/groupby.py:3043]: idxmax/idxmin need to default to None for backwards compatibility; using _resolve_numeric_only would currently have them default to True. The different being that currently idxmin/idxmax will operate on categorical columns with numeric_only being None; with numeric_only True they would be dropped.

I do plan on making all necessary methods go through _resolve_numeric_only as part of #46072, including idxmin/idxmax. The logic would there would need to be made method-specific, and was going to do all methods together.

File: pandas/core/groupby/groupby.py (hunk 3 above)
Comment (rhshadrach) [reply pandas/core/groupby/groupby.py]: Thanks - I was able to remove this hack by setting the appropriate value of raise_on_typeerror.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @rhshadrach very nice!

**Pull request merged**